subsys/bluetooth/mesh/scheduler_srv.c

```diff
@@ -0,0 +1,630 @@
+				return false;
+			}
+			month = item->month;
+		}
+	}
+
+	sched_time->tm_mon = __builtin_ctz(month);
+	return true;
+}
+
+static bool set_day(struct tm *sched_time,
+		    struct tm *current_local,
+		    struct bt_mesh_schedule_register *item,
+		    struct bt_mesh_time_srv *srv)
+{
+	if (item->day < current_local->tm_mday && item->day != ANY_DAY) {
+		return false;
+	}
+
+	if (!item->day_of_week) {
+		return false;
+	}
+
+	sched_time->tm_mday = item->day == ANY_DAY ?
+			current_local->tm_mday : item->day;
+
+	sched_time = bt_mesh_time_srv_localtime(srv,
+			bt_mesh_time_srv_mktime(srv, sched_time));
+	if (item->day_of_week & (1 << sched_time->tm_wday)) {
+		return true;
+	}
+
+	if (item->day == ANY_DAY) {
+		int rest_wday = item->day_of_week >> sched_time->tm_wday;
+		int delta = rest_wday ? __builtin_ctz(rest_wday) :
+			__builtin_ctz(item->day_of_week) +
+			6 - sched_time->tm_wday;
+		return revise_month(sched_time, current_local, item, delta);
+	}
+
+	return false;
+}
+
+static bool set_hour(struct tm *sched_time,
+		     struct tm *current_local,
+		     struct bt_mesh_schedule_register *item,
+		     struct bt_mesh_time_srv *srv)
+{
+	if (item->hour < current_local->tm_hour) {
+		return false;
+	}
+
+	if (item->hour == RANDOM_HOUR) {
+		sched_time->tm_hour = sys_rand32_get() % 24;
+		return revise_day(sched_time, current_local, item, srv, 1);
+	}
+
+	sched_time->tm_hour = item->hour == ANY_HOUR ?
+			current_local->tm_hour : item->hour;
+
+	return true;
+}
+
+static bool set_minute(struct tm *sched_time,
+		       struct tm *current_local,
+		       struct bt_mesh_schedule_register *item,
+		       struct bt_mesh_time_srv *srv)
+{
+	if (item->minute < current_local->tm_min) {
+		return false;
+	}
+
+	bool ovflw = false;
+
+	if (item->minute == EVERY_15_UNITS) {
+		sched_time->tm_min = current_local->tm_min < 15 ? 15 :
+				current_local->tm_min < 30 ? 30 :
+				current_local->tm_min < 45 ? 45 : 0;
+		ovflw = current_local->tm_min == 0 ? true : false;
+	} else if (item->minute == EVERY_20_UNITS) {
+		sched_time->tm_min = current_local->tm_min < 20 ? 20 :
+				current_local->tm_min < 40 ? 40 : 0;
+		ovflw = current_local->tm_min == 0 ? true : false;
+	} else if (item->minute == RANDOM_UNIT) {
+		sched_time->tm_min = sys_rand32_get() % 60;
+		ovflw = true;
+	} else {
+		sched_time->tm_min = item->minute == ANY_UNIT ?
+				current_local->tm_min : item->minute;
+	}
+
+	if (ovflw) {
+		return revise_hour(sched_time, current_local, item, srv, 1);
+	}
+
+	return true;
+}
+
+static bool set_second(struct tm *sched_time,
+		       struct tm *current_local,
+		       struct bt_mesh_schedule_register *item,
+		       struct bt_mesh_time_srv *srv)
+{
+	if (item->second < current_local->tm_sec) {
+		return false;
+	}
+
+	bool ovflw = false;
+
+	if (item->second == EVERY_15_UNITS) {
+		sched_time->tm_sec = current_local->tm_sec < 15 ? 15 :
+				current_local->tm_sec < 30 ? 30 :
+				current_local->tm_sec < 45 ? 45 : 0;
```

Comment:
`sched_time->tm_sec = 15 * ceiling_fraction(current_local->tm_sec, 15)`

Reply:
Nop. It gives a different result to what I have here. Just quickly checked in the online compiler. Values from 0 to 60.